qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,215 @@
+
+    """
+
+    result = [l]
+    for i, x in enumerate(l):
+        if len(list(x[1])) > 1:
+            members = list(x[1])
+            temp = copy.deepcopy(result)
+
+            # update result with first element of types in member
+            for each_list in result:
+                each_list[i][1] = members[0]
+
+            # add in other permutations of types in member
+            for n in range(1, len(members)):
+                copy_result = copy.deepcopy(temp)
+                for each_list in copy_result:
+                    each_list[i][1] = members[n]
+                result += copy_result
+    return result
+
+
+def generate_nodes_by_action(action, opt):
+    """
+    Given a method, generates all combinations of inputs and
+    outputs for that particular method and and stores the combinations
+    as dictionaries in a resulting list.
+
+    Parameters
+    ----------
+    method : Qiime2.action
+    opt : {True, False}
+        Delineates if optional types should be included
+
+    Returns
+    -------
+    list of dictionaries - each dictionary is a combination inputs and
+        outputs for particular node
+
+    """
+
+    input, input_nr = get_next_arguments(action, "input")
+    param, param_nr = get_next_arguments(action, "param")
+    output, output_nr = get_next_arguments(action, "output")
+
+    input = unravel(input)
+    param = unravel(param)
+
+    opt_in_list = []
+
+    if opt:
+        opt_in_list += input_nr
+        opt_in_list += param_nr
+        opt_in_list = unravel(opt_in_list)
+        ins = [dict(x) for x in
+               [list(chain.from_iterable(i)) for i in
+                list(product(input, param, opt_in_list))]]
+        outs = dict(output + output_nr)
+        results = [{'inputs': i, 'outputs': outs} for i in ins]
+        return results
+
+    ins = [dict(x) for x in
+           [list(chain.from_iterable(i)) for i in list(product(input, param))]]
+    outs = dict(output)
+    results = [{'inputs': i, 'outputs': outs} for i in ins]
+    return results
+
+
+def build_graph(sigs=[], opt=False):
+    """
+    Constructs a networkx graph with different semantic types
+    and actions as nodes
+
+    Parameters
+    ----------
+    sigs : list of Qiime2.action
+    opt : {True, False}
+        Delineates if optional types should be included in the graph
+
+    Returns
+    -------
+    nx.DiGraph - networkx graph connected based on all or specified methods
+
+    """
+
+    G = nx.DiGraph()
+    G.edges(data=True)
+    action_list = []
+    sigs = [x.__name__ for x in sigs]
+
+    # get all actions or specifc actions if specified in sigs
+    pm = qiime2.sdk.PluginManager()
+    if not sigs:
+        for _, pg in pm.plugins.items():
+            action_list += list(pg.actions.values())
+    else:
+        for _, pg in pm.plugins.items():
+            for action in pg.actions.keys():
+                if action in sigs:
+                    action_list.append(pg.actions[str(action)])
```

Comment:
This block can become `action_list = sigs` (also, it might be good to call `sigs` just `action_list` instead in the `build_graph` signature).

Reply:
So passing in methods and actions into sigs, are we passing in the <method > or <visualizer >  object that has the action.signature.inputs and action.signature.parameters or are we passing in a list of <functions> that is an object of the function. 

In the later case, we would need to search the plugin for the <method > or <visualizer > object to find the action's attributes.

Reply:
You would get the actual action object, so you'd have a list of things that have the `.signature` property ready to go